server/findproviders.go

```diff
@@ -1,98 +1,132 @@
 package server
 
 import (
-	"bytes"
-	"encoding/json"
+	"context"
 	"net/http"
 
 	"github.com/ipfs/go-cid"
 	"github.com/ipfs/go-delegated-routing/client"
-	"github.com/ipfs/go-delegated-routing/parser"
+	proto "github.com/ipfs/go-delegated-routing/gen/proto"
 	logging "github.com/ipfs/go-log"
+	"github.com/ipld/edelweiss/values"
 	"github.com/libp2p/go-libp2p-core/peer"
 	"github.com/multiformats/go-multiaddr"
 )
 
-var log = logging.Logger("delegated/server")
+var logger = logging.Logger("service/server/DelegatedRouting")
 
-type FindProvidersAsyncFunc func(cid.Cid, chan<- client.FindProvidersAsyncResult) error
+type DelegatedRoutingService interface {
+	FindProviders(key cid.Cid, rch chan<- client.FindProvidersAsyncResult) error
+	GetIPNS(id []byte, rch chan<- client.GetIPNSAsyncResult) error
+	PutIPNS(id []byte, record []byte, rch chan<- client.PutIPNSAsyncResult) error
+}
```

Comment:
Both conventions would work for the typical use case. However, this convention provides strictly more flexibility to the caller of the interface. E.g. it allows the caller to reuse existing channels instead of creating new ones in tricker applications.

Reply:
> E.g. it allows the caller to reuse existing channels instead of creating new ones in trickier applications.

Does it? If the interface declares that the channel is closed by the functions (e.g. `FindProviders`) then the channel can't be reused. If the interface declares that the channel is closed by the caller then there's no way for the caller to know when it should close the channel.

What were you envisioning here?